hugr/src/types/type_param.rs

```diff
@@ -252,15 +270,36 @@ impl TypeArg {
             TypeArg::Extensions { es: _ } => Ok(()),
             TypeArg::Variable {
                 v: TypeArgVariable { idx, cached_decl },
-            } => check_typevar_decl(var_decls, *idx, cached_decl),
+            } => {
+                assert!(
+                    match cached_decl {
+                        TypeParam::Type { .. } => false,
+                        TypeParam::List { param } if matches!(**param, TypeParam::Type { .. }) =>
+                            false,
+                        _ => true,
+                    },
+                    "Malformed TypeArg::Variable {} - should be inconstructible",
+                    cached_decl
+                );
+
+                check_typevar_decl(var_decls, *idx, cached_decl)
+            }
         }
     }
 
     pub(crate) fn substitute(&self, t: &Substitution) -> Self {
         match self {
-            TypeArg::Type { ty } => TypeArg::Type {
-                ty: ty.substitute(t),
-            },
+            TypeArg::Type { ty } => {
+                // A row variable standing for many types is represented as a single type
+                // TODO: this case can't happen until we start substituting across Hugrs
```

Comment:
I'm not sure this "can't happen until". But I haven't come up with a case yet. While *TypeDef*s do not allow row vars (only *OpDef*s), *and* we haven't tackled substitution across nodes (#709) it may indeed be that it can't happen. But, I've captured it in a unit test, and the impl here was broken, so fixed.